README
-*- coding: utf-8-unix -*-

				CCMemory
				========


Topics
------

  1. Introduction
  2. License
  3. Install
  4. Usage
  A. Credits
  B. Bugs, vulnerabilities and contributions
  C. Resources
  D. Badges and static analysis


1. Introduction
---------------

This package installs  a C11 language library  implementing an interface
for custom memory allocators.

  This package  supports POSIX platforms.   This package depends  on the
package CCExceptions, at  least version 0.10.0.  To run  the test suite:
this package depends upon the package CCTests, at least version 0.5.0.

  The package uses the GNU Autotools  and it is tested, using Travis CI,
on both Ubuntu GNU+Linux systems and OS X systems.


2. License
----------

Copyright (c) 2016, 2018-2020 Marco Maggi <mrc.mgg@gmail.com>

This program is free software: you  can redistribute it and/or modify it
under the terms of the GNU Lesser General Public License as published by
the Free  Software Foundation, either version  3 of the License,  or (at
your option) any later version.

This program  is distributed  in the  hope that it  will be  useful, but
WITHOUT   ANY   WARRANTY;  without   even   the   implied  warranty   of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser
General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.


3. Install
----------

To install from a proper release tarball, do this:

    $ cd ccmemory-0.1.0
    $ mkdir build
    $ cd build
    $ ../configure
    $ make
    $ make check
    $ make install

to inspect the available configuration options:

    $ ../configure --help

  The Makefile is designed to allow parallel builds, so we can do:

   $ make -j4 all && make -j4 check

which,  on  a  4-core  CPU,   should  speed  up  building  and  checking
significantly.

  The  Makefile supports  the  DESTDIR environment  variable to  install
files in a temporary location, example: to see what will happen:

    $ make -n install DESTDIR=/tmp/ccmemory

to really do it:

    $ make install DESTDIR=/tmp/ccmemory

  After the installation it is  possible to verify the installed library
against the test suite with:

    $ make installcheck

  From  a repository  checkout or  snapshot  (the ones  from the  GitHub
site): we  must install the  GNU Autotools (GNU Automake,  GNU Autoconf,
GNU Libtool),  then we must first  run the script "autogen.sh"  from the
top source directory, to generate the needed files:

    $ cd ccmemory
    $ sh autogen.sh

notice  that  "autogen.sh"  will   run  the  programs  "autoreconf"  and
"libtoolize"; the  latter is  selected through the  environment variable
"LIBTOOLIZE",  whose  value  can  be  customised;  for  example  to  run
"glibtoolize" rather than "libtoolize" we do:

   $ LIBTOOLIZE=glibtoolize sh autogen.sh

  After this the  procedure is the same  as the one for  building from a
proper release tarball, but we have to enable maintainer mode:

    $ ../configure --enable-maintainer-mode [options]
    $ make
    $ make check
    $ make install


4. Usage
--------

Read the documentation generated from  the Texinfo sources.  The package
installs the documentation  in Info format; we can  generate and install
documentation in HTML format by running:

   $ make html
   $ make install-html


A. Credits
----------

The  stuff was  written by  Marco Maggi.   If this  package exists  it's
because of the great GNU software tools that he uses all the time.


B. Bugs, vulnerabilities and contributions
------------------------------------------

Bug  and vulnerability  reports are  appreciated, all  the vulnerability
reports  are  public; register  them  using  the  Issue Tracker  at  the
project's GitHub  site.  For  contributions and  patches please  use the
Pull Requests feature at the project's GitHub site.


C. Resources
------------

The latest release of this package can be downloaded from:

	 <https://bitbucket.org/marcomaggi/ccmemory/downloads>

development takes place at:

	       <https://github.com/marcomaggi/ccmemory/>

and as backup at:

	      <https://bitbucket.org/marcomaggi/ccmemory/>

the documentation is available online:

	   <https://marcomaggi.github.io/docs/ccmemory.html>

the GNU Project software can be found here:

			 <https://www.gnu.org/>

the package CCExceptions is available at:

	     <https://github.com/marcomaggi/ccexceptions/>

the package CCTests is available at:

		<https://github.com/marcomaggi/cctests/>


D. Badges and static analysis
-----------------------------

D.1 Travis CI
-------------

Travis CI is  a hosted, distributed continuous  integration service used
to build and test software projects  hosted at GitHub.  We can find this
project's dashboard at:

	      <https://travis-ci.org/marcomaggi/ccmemory>

  Usage of this service is configured through the file ".travis.yml" and
the scripts under the directory "meta/travis-ci".


D.2 Clang's Static Analyzer
---------------------------

The Clang Static Analyzer is a source code analysis tool that finds bugs
in C, C++, and Objective-C programs.  It is distributed along with Clang
and we can find it at:

		   <http://clang-analyzer.llvm.org/>

  Usage of this service is implemented with make rules; see the relevant
section in the file "Makefile.am".


D.3 Codecov
-----------

Codecov is a service providing code  coverage reports.  We can find this
project's dashboard at:

	      <https://codecov.io/gh/marcomaggi/ccmemory>

Usage of  this service is  implemented through direct  interface between
GitHub and Codecov  sites; it configured through  the file `codecov.yml`
and appropriate entries in Travis CI's matrix of builds.


D.4 Codacy
----------

Codacy is  an online service  providing code  review.  We can  find this
project's dashboard at:

	    <https://www.codacy.com/app/marcomaggi/ccmemory>

Usage of this service is  implemented through direct integration between
GitHub and Codacy sites.


### end of file
# Local Variables:
# mode: text
# fill-column: 72
# End:


NEWS
-*- coding: utf-8-unix -*-

			News file for CCMemory
			======================

Revision 0.3.0-devel.0
----------------------

* Upgraded to CCExceptions version 0.10.0.

* Upgraded to CCTests version 0.5.0.


Revision 0.2.2-devel.2
----------------------

* Packaging infrastructure fixes.


Revision 0.2.2-devel.1
----------------------

* Development of internal struct accessors.


Revision 0.2.2-devel.0
----------------------

* Upgraded dependency upon CCExceptions to version 0.9.0.

* Upgraded dependency upon CCTests to version 0.4.1.


Revision 0.2.1-devel.4
----------------------

* Package infrastructure development.


Revision 0.2.1-devel.3
----------------------

* Fixed documentation of some functions using the standard allocator.

* Package infrastructure development.


Revision 0.2.1-devel.2
----------------------

* Development of build infrastructure.


Revision 0.2.1-devel.1
----------------------

* Development of ASCII and ASCIIZ string representations.

* BACKWARDS INCOMPATIBILITY Total review of memory blocks module.


Revision 0.2.1-devel.0
----------------------

* Added exception handlers and guarded memory allocators.


Revision 0.2.0-devel.1
----------------------

* Updated to CCExceptions v0.8.0-devel.4.

* Added missing C headers includes.


Revision 0.2.0-devel.0
----------------------

* Changed  API:  the  allocator's  free function  no  longer  raises  an
  exception.


Revision 0.1.0-devel.3
----------------------

* Added  missing URLs  for  dependencies to  the  documentation and  the
  README files.

* Updated to CCExceptions v0.8.0-devel.2 and CCTests v0.3.0-devel.0.


Revision 0.1.0-devel.2
----------------------

* Better check for CCTests in "configure.ac".


Revision 0.1.0-devel.1
----------------------

* Fixed data script for "pkg-config".


Revision 0.1.0-devel.0
----------------------

* Initial development release.


### end of file
# Local Variables:
# mode: text
# fill-column: 72
# paragraph-start: "*"
# ispell-local-dictionary: "en_GB"
# End:


tests/version.c
/*
  Part of: CCMemory
  Contents: test for version functions
  Date: Sat Dec 17, 2016

  Abstract

	Test file for version functions.

  Copyright (C) 2016, 2018 Marco Maggi <mrc.mgg@gmail.com>

  See the COPYING file.
*/

#include <ccmemory.h>
#include <stdio.h>
#include <stdlib.h>

int
main (void)
{
  printf("version number string: %s\n", ccmem_version_string());
  printf("libtool version number: %d:%d:%d\n",
	 ccmem_version_interface_current(),
	 ccmem_version_interface_revision(),
	 ccmem_version_interface_age());
  exit(EXIT_SUCCESS);
}

/* end of file */


tests/ascii-strings.c
/*
  Part of: CCMemory
  Contents: tests for ASCII strings
  Date: Sep 26, 2018

  Abstract



  Copyright (C) 2018 Marco Maggi <mrc.mgg@gmail.com>

  This is free software; you  can redistribute it and/or modify it under
  the terms of the GNU Lesser General Public License as published by the
  Free Software  Foundation; either version  3.0 of the License,  or (at
  your option) any later version.

  This library  is distributed in the  hope that it will  be useful, but
  WITHOUT   ANY  WARRANTY;   without  even   the  implied   warranty  of
  MERCHANTABILITY  or FITNESS  FOR A  PARTICULAR PURPOSE.   See  the GNU
  Lesser General Public License for more details.

  You  should have  received a  copy of  the GNU  Lesser  General Public
  License along  with this library; if  not, write to  the Free Software
  Foundation, Inc.,  59 Temple Place,  Suite 330, Boston,  MA 02111-1307
  USA.

*/


/** --------------------------------------------------------------------
 ** Headers.
 ** ----------------------------------------------------------------- */

#include "ccmemory.h"
#include <cctests.h>


/** --------------------------------------------------------------------
 ** Constructors.
 ** ----------------------------------------------------------------- */

void
test_1_1_1 (cce_destination_t L)
/* Test for "ccmem_new_ascii()". */
{
  static char *		str = "ciao";
  size_t 		len = strlen(str);
  ccmem_ascii_t		S   = ccmem_new_ascii(str, len);

  cctests_assert_ascii(L, str, S.ptr, S.len);
  cctests_assert_equal_size(L, len, S.len);
}

void
test_1_2_1 (cce_destination_t L)
/* Test for "ccmem_new_ascii_empty()". */
{
  static char *		str = "";
  ccmem_ascii_t		S   = ccmem_new_ascii_empty();

  cctests_assert_ascii(L, str, S.ptr, S.len);
  cctests_assert_equal_size(L, 0, S.len);
}

void
test_1_3_1 (cce_destination_t L)
/* Test for "ccmem_new_ascii_null()". */
{
  ccmem_ascii_t		S   = ccmem_new_ascii_null();

  cctests_assert_equal_pointer(L, NULL, S.ptr);
  cctests_assert_equal_size(L, 0, S.len);
}

void
test_1_4_1 (cce_destination_t L)
/* Test for "ccmem_new_ascii_from_str()". */
{
  static char *		str = "ciao";
  size_t 		len = strlen(str);
  ccmem_ascii_t		S = ccmem_new_ascii_from_str(str);

  cctests_assert_ascii(L, str, S.ptr, S.len);
  cctests_assert_equal_size(L, len, S.len);
}

void
test_1_5_1 (cce_destination_t L)
/* Test for "ccmem_new_ascii_from_block()". */
{
  static char *		str = "ciao";
  size_t 		len = strlen(str);
  ccmem_block_t		B   = ccmem_new_block((uint8_t *)str, len);
  ccmem_ascii_t		S   = ccmem_new_ascii_from_block(B);

  cctests_assert_equal_pointer(L, B.ptr, S.ptr);
  cctests_assert_equal_size(L, B.len, S.len);
}

void
test_1_6_1 (cce_destination_t L)
/* Test for "ccmem_new_ascii_from_asciiz()". */
{
  static char *		str = "ciao";
  size_t 		len = strlen(str);
  ccmem_asciiz_t	B   = ccmem_new_asciiz(str, len);
  ccmem_ascii_t		S   = ccmem_new_ascii_from_asciiz(B);

  cctests_assert_equal_pointer(L, B.ptr, S.ptr);
  cctests_assert_equal_size(L, B.len, S.len);
}


/** --------------------------------------------------------------------
 ** Predicates.
 ** ----------------------------------------------------------------- */

void
test_2_1_1 (cce_destination_t L)
/* Test for "ccmem_ascii_is_empty()". */
{
  static char *		str = "ciao";
  size_t 		len = strlen(str);
  ccmem_ascii_t		S   = ccmem_new_ascii(str, len);

  cctests_assert(L, ! ccmem_ascii_is_empty(S));
}

void
test_2_1_2 (cce_destination_t L)
/* Test for "ccmem_ascii_is_empty()". */
{
  ccmem_ascii_t		S = ccmem_new_ascii_empty();

  cctests_assert(L, ccmem_ascii_is_empty(S));
}

void
test_2_1_3 (cce_destination_t L)
/* Test for "ccmem_ascii_is_empty()". */
{
  ccmem_ascii_t		S = ccmem_new_ascii_null();

  cctests_assert(L, ! ccmem_ascii_is_empty(S));
}

/* ------------------------------------------------------------------ */

void
test_2_2_1 (cce_destination_t L)
/* Test for "ccmem_ascii_is_null()". */
{
  static char *		str = "ciao";
  size_t 		len = strlen(str);
  ccmem_ascii_t		S   = ccmem_new_ascii(str, len);

  cctests_assert(L, ! ccmem_ascii_is_null(S));
}

void
test_2_2_2 (cce_destination_t L)
/* Test for "ccmem_ascii_is_null()". */
{
  ccmem_ascii_t		S   = ccmem_new_ascii_empty();

  cctests_assert(L, ! ccmem_ascii_is_null(S));
}

void
test_2_2_3 (cce_destination_t L)
/* Test for "ccmem_ascii_is_null()". */
{
  ccmem_ascii_t		S = ccmem_new_ascii_null();

  cctests_assert(L, ccmem_ascii_is_null(S));
}


/** --------------------------------------------------------------------
 ** Comparison.
 ** ----------------------------------------------------------------- */

void
test_3_1_1 (cce_destination_t L)
/* Test for "ccmem_ascii_equal()". */
{
  static char *		str = "ciao";
  size_t 		len = strlen(str);
  ccmem_ascii_t		S1  = ccmem_new_ascii(str, len);
  ccmem_ascii_t		S2  = ccmem_new_ascii(str, len);

  cctests_assert(L, ccmem_ascii_equal(S1, S2));
}

void
test_3_1_2 (cce_destination_t L)
/* Test for "ccmem_ascii_equal()". */
{
  static char *		str = "ciao";
  size_t 		len = strlen(str);
  ccmem_ascii_t		S1  = ccmem_new_ascii(str, len);
  ccmem_ascii_t		S2  = ccmem_new_ascii(str, len-1);

  cctests_assert(L, ! ccmem_ascii_equal(S1, S2));
}

void
test_3_1_3 (cce_destination_t L)
/* Test for "ccmem_ascii_equal()". */
{
  static char *		str = "ciao";
  size_t 		len = strlen(str);
  ccmem_ascii_t		S1  = ccmem_new_ascii(str, len-1);
  ccmem_ascii_t		S2  = ccmem_new_ascii(str, len);

  cctests_assert(L, ! ccmem_ascii_equal(S1, S2));
}

void
test_3_1_4 (cce_destination_t L)
/* Test for "ccmem_ascii_equal()". */
{
  static char *		str1 = "ciao";
  static char *		str2 = "hey!";
  size_t 		len1 = strlen(str1);
  size_t 		len2 = strlen(str2);
  ccmem_ascii_t		S1  = ccmem_new_ascii(str1, len1);
  ccmem_ascii_t		S2  = ccmem_new_ascii(str2, len2);

  cctests_assert(L, ! ccmem_ascii_equal(S1, S2));
}


/** --------------------------------------------------------------------
 ** Operations.
 ** ----------------------------------------------------------------- */

void
test_4_1_1 (cce_destination_t L)
/* Test for "ccmem_ascii_clean_memory()". */
{
  size_t 		len     = 4;
  char			str1[4] = { 0x01, 0x02, 0x03, 0x04 };
  char			str2[4] = { 0x00, 0x00, 0x00, 0x00 };
  ccmem_ascii_t		S       = ccmem_new_ascii(str1, len);

  ccmem_ascii_clean_memory(S);
  cctests_assert(L, 0 == memcmp(str2, S.ptr, S.len));
}


/** --------------------------------------------------------------------
 ** Dynamic memory allocation.
 ** ----------------------------------------------------------------- */

void
test_5_1_1 (cce_destination_t upper_L)
/* Test for "ccmem_ascii_malloc()". */
{
  cce_location_t	L[1];

  if (cce_location(L)) {
    cce_run_catch_handlers_raise(L, upper_L);
  } else {
    ccmem_ascii_t	S = ccmem_ascii_malloc(L, ccmem_standard_allocator, 4);

    S.ptr[0] = 'c';
    S.ptr[1] = 'i';
    S.ptr[2] = 'a';
    S.ptr[3] = 'o';

    cctests_assert_equal_size(L, 4, S.len);
    cctests_assert_ascii(L, "ciao", S.ptr, S.len);

    ccmem_ascii_free(ccmem_standard_allocator, S);
    cce_run_body_handlers(L);
  }
}

void
test_5_2_1 (cce_destination_t upper_L)
/* Test for "ccmem_ascii_realloc()". */
{
  cce_location_t	L[1];

  if (cce_location(L)) {
    cce_run_catch_handlers_raise(L, upper_L);
  } else {
    ccmem_ascii_t	S = ccmem_ascii_malloc(L, ccmem_standard_allocator, 4);

    S.ptr[0] = 'c';
    S.ptr[1] = 'i';
    S.ptr[2] = 'a';
    S.ptr[3] = 'o';

    cctests_assert_equal_size(L, 4, S.len);
    cctests_assert_ascii(L, "ciao", S.ptr, S.len);

    S = ccmem_ascii_realloc(L, ccmem_standard_allocator, S, 4+6);

    S.ptr[4] = ' ';
    S.ptr[5] = 'm';
    S.ptr[6] = 'a';
    S.ptr[7] = 'm';
    S.ptr[8] = 'm';
    S.ptr[9] = 'a';

    cctests_assert_equal_size(L, 4+6, S.len);
    cctests_assert_ascii(L, "ciao mamma", S.ptr, S.len);

    ccmem_ascii_free(ccmem_standard_allocator, S);
    cce_run_body_handlers(L);
  }
}


/** --------------------------------------------------------------------
 ** Guarded memory allocation.
 ** ----------------------------------------------------------------- */

void
test_6_1_1 (cce_destination_t upper_L)
/* Test for "ccmem_ascii_malloc_guardeda()". */
{
  cce_location_t	L[1];
  ccmem_clean_handler_t	S_H[1];

  if (cce_location(L)) {
    cce_run_catch_handlers_raise(L, upper_L);
  } else {
    ccmem_ascii_t	S = ccmem_ascii_malloc_guarded(L, S_H, ccmem_standard_allocator, 4);

    S.ptr[0] = 'c';
    S.ptr[1] = 'i';
    S.ptr[2] = 'a';
    S.ptr[3] = 'o';

    cctests_assert_equal_size(L, 4, S.len);
    cctests_assert_ascii(L, "ciao", S.ptr, S.len);
    cce_run_body_handlers(L);
  }
}

void
test_6_2_1 (cce_destination_t upper_L)
/* Test for "ccmem_ascii_realloc_guarded()". */
{
  cce_location_t	L[1];
  ccmem_clean_handler_t	S_H[1];

  if (cce_location(L)) {
    cce_run_catch_handlers_raise(L, upper_L);
  } else {
    ccmem_ascii_t	S = ccmem_ascii_malloc_guarded(L, S_H, ccmem_standard_allocator, 4);

    S.ptr[0] = 'c';
    S.ptr[1] = 'i';
    S.ptr[2] = 'a';
    S.ptr[3] = 'o';

    cctests_assert_equal_size(L, 4, S.len);
    cctests_assert_ascii(L, "ciao", S.ptr, S.len);

    S = ccmem_ascii_realloc_guarded(L, S_H, ccmem_standard_allocator, S, 4+6);

    S.ptr[4] = ' ';
    S.ptr[5] = 'm';
    S.ptr[6] = 'a';
    S.ptr[7] = 'm';
    S.ptr[8] = 'm';
    S.ptr[9] = 'a';

    cctests_assert_equal_size(L, 4+6, S.len);
    cctests_assert_ascii(L, "ciao mamma", S.ptr, S.len);
    cce_run_body_handlers(L);
  }
}


int
main (void)
{
  cctests_init("tests for ASCII strings");
  {
    cctests_begin_group("constructors");
    {
      cctests_run(test_1_1_1);
      cctests_run(test_1_2_1);
      cctests_run(test_1_3_1);
      cctests_run(test_1_4_1);
      cctests_run(test_1_5_1);
    }
    cctests_end_group();

    cctests_begin_group("predicates");
    {
      cctests_run(test_2_1_1);
      cctests_run(test_2_1_2);
      cctests_run(test_2_1_3);
      cctests_run(test_2_2_1);
      cctests_run(test_2_2_2);
      cctests_run(test_2_2_3);
    }
    cctests_end_group();

    cctests_begin_group("comparison");
    {
      cctests_run(test_3_1_1);
      cctests_run(test_3_1_2);
      cctests_run(test_3_1_3);
      cctests_run(test_3_1_4);
    }
    cctests_end_group();

    cctests_begin_group("operations");
    {
      cctests_run(test_4_1_1);
    }

    cctests_begin_group("dynamic memory allocation");
    {
      cctests_run(test_5_1_1);
      cctests_run(test_5_2_1);
    }

    cctests_begin_group("guarded memory allocation");
    {
      cctests_run(test_6_1_1);
      cctests_run(test_6_2_1);
    }
    cctests_end_group();
  }
  cctests_final();
}

/* end of file */


src/ccmemory-blocks.c
/*
  Part of: CCMemory
  Contents: functions for memory blocks
  Date: May  8, 2018

  Abstract


  Copyright (C) 2018, 2020 Marco Maggi <mrc.mgg@gmail.com>

  This is free software; you  can redistribute it and/or modify it under
  the terms of the GNU Lesser General Public License as published by the
  Free Software  Foundation; either version  3.0 of the License,  or (at
  your option) any later version.

  This library  is distributed in the  hope that it will  be useful, but
  WITHOUT   ANY  WARRANTY;   without  even   the  implied   warranty  of
  MERCHANTABILITY  or FITNESS  FOR A  PARTICULAR PURPOSE.   See  the GNU
  Lesser General Public License for more details.

  You  should have  received a  copy of  the GNU  Lesser  General Public
  License along  with this library; if  not, write to  the Free Software
  Foundation, Inc.,  59 Temple Place,  Suite 330, Boston,  MA 02111-1307
  USA.
*/


/** --------------------------------------------------------------------
 ** Headers.
 ** ----------------------------------------------------------------- */

#include "ccmemory-internals.h"
#include <stdio.h>


/** --------------------------------------------------------------------
 ** Operations.
 ** ----------------------------------------------------------------- */

ccmem_block_t
ccmem_block_shift (ccmem_block_t const A, ssize_t const offset, size_t const dim)
{
  if ((0 == dim) || (0 == offset)) {
    return A;
  } else if (1 == dim) {
    ccmem_block_t	B = {
      .ptr = A.ptr + offset,
      .len = A.len + offset
    };
    return B;
  } else if (1 == offset) {
    ccmem_block_t	B = {
      .ptr = A.ptr + dim,
      .len = A.len + dim
    };
    return B;
  } else {
    ccmem_block_t	B = {
      .ptr = A.ptr + offset * dim,
      .len = A.len + offset * dim
    };
    return B;
  }
}

ccmem_block_t
ccmem_block_difference (ccmem_block_t A, ccmem_block_t B)
{
  assert(A.ptr == B.ptr);
  assert(A.len >= B.len);
  ccmem_block_t	C = {
    .ptr = A.ptr + B.len,
    .len = A.len - B.len
  };
  return C;
}

/* end of file */


src/ccmemory-version.c
/*
  Part of: CCMemory
  Contents: version functions
  Date: Dec 17, 2016

  Abstract



  Copyright (C) 2016, 2018 Marco Maggi <mrc.mgg@gmail.com>

  This program is free  software: you can redistribute it and/or  modify it under the
  terms of the  GNU Lesser General Public  License as published by  the Free Software
  Foundation, either version 3 of the License, or (at your option) any later version.

  This program  is distributed in the  hope that it  will be useful, but  WITHOUT ANY
  WARRANTY; without  even the implied  warranty of  MERCHANTABILITY or FITNESS  FOR A
  PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License along with
  this program.  If not, see <http://www.gnu.org/licenses/>.
*/


/** --------------------------------------------------------------------
 ** Headers.
 ** ----------------------------------------------------------------- */

#include "ccmemory-internals.h"


const char *
ccmem_version_string (void)
{
  return ccmemory_VERSION_INTERFACE_STRING;
}
int
ccmem_version_interface_current (void)
{
  return ccmemory_VERSION_INTERFACE_CURRENT;
}
int
ccmem_version_interface_revision (void)
{
  return ccmemory_VERSION_INTERFACE_REVISION;
}
int
ccmem_version_interface_age (void)
{
  return ccmemory_VERSION_INTERFACE_AGE;
}

/* end of file */


src/ccmemory-internals.h
/*
  Part of: CCMemory
  Contents: private header file
  Date: Dec 17, 2016

  Abstract

	This  header  file is  for  internal  definitions.  It  must  be
	included by all the source files in this package.

  Copyright (C) 2016, 2018 Marco Maggi <mrc.mgg@gmail.com>

  This program is  free software: you can redistribute  it and/or modify
  it  under the  terms  of  the GNU  Lesser  General  Public License  as
  published by  the Free  Software Foundation, either  version 3  of the
  License, or (at your option) any later version.

  This program  is distributed in the  hope that it will  be useful, but
  WITHOUT   ANY  WARRANTY;   without  even   the  implied   warranty  of
  MERCHANTABILITY  or FITNESS  FOR A  PARTICULAR PURPOSE.   See  the GNU
  General Public License for more details.

  You  should have  received a  copy of  the GNU  Lesser General  Public
  License     along    with     this    program.      If    not,     see
  <http://www.gnu.org/licenses/>.
*/

#ifndef CCMEMORY_INTERNALS_H
#define CCMEMORY_INTERNALS_H 1


/** --------------------------------------------------------------------
 ** Headers.
 ** ----------------------------------------------------------------- */

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif
#include "ccmemory.h"
#include <assert.h>


/** --------------------------------------------------------------------
 ** Preprocessor macros.
 ** ----------------------------------------------------------------- */



/** --------------------------------------------------------------------
 ** Constants.
 ** ----------------------------------------------------------------- */




/** --------------------------------------------------------------------
 ** Function prototypes.
 ** ----------------------------------------------------------------- */



/** --------------------------------------------------------------------
 ** Done.
 ** ----------------------------------------------------------------- */

#ifdef __cplusplus
} // extern "C"
#endif

#endif /* CCMEMORY_INTERNALS_H */

/* end of file */


meta/travis-ci/install-cctests.sh
#!/bin/bash
#
# Installation  script to  run from  the  Travis CI  config file  before
# attempting a build.
#
# Install CCTests under the directory "/tmp/mine".  We assume the script
# is run from the top directory of the build tree.

PROGNAME=install-cctests.sh
VERSION=0.3.0-devel.3
STEM="cctests-${VERSION}"
ARCHIVE="${STEM}.tar.gz"
SOURCE_URI="https://github.com/marcomaggi/cctests/archive/v${VERSION}.tar.gz"
LOCAL_ARCHIVE="/tmp/${ARCHIVE}"
TOP_SRCDIR="/tmp/${STEM}"

test -d /tmp/mine || {
    if ! mkdir /tmp/mine
    then
	printf '%s: error creating directory for dependency package building and installation\n' "$PROGNAME" >&2
	exit 1
    fi
}

echo "wget \"$SOURCE_URI\" -O \"$LOCAL_ARCHIVE\"" >&2
if ! wget "$SOURCE_URI" -O "$LOCAL_ARCHIVE"
then
    printf '%s: error downloading %s\n' "$PROGNAME" "${ARCHIVE}" >&2
    exit 1
fi

cd /tmp

echo "tar -xzf \"$LOCAL_ARCHIVE\"" >&2
if ! tar -xzf "$LOCAL_ARCHIVE"
then
    printf '%s: error unpacking %s\n' "$PROGNAME" "$LOCAL_ARCHIVE" >&2
    exit 1
fi

if ! cd "$TOP_SRCDIR"
then
    printf '%s: error changing directory to %s\n' "$PROGNAME" "${TOP_SRCDIR}" >&2
    exit 1
fi

echo "sh autogen.sh" >&2
if [[ "$TRAVIS_OS_NAME" == "osx" ]]
then
    if ! LIBTOOLIZE=glibtoolize sh autogen.sh
    then
	printf '%s: error configuring %s\n' "$PROGNAME" "${STEM}" >&2
	exit 1
    fi
else
    if ! sh autogen.sh
    then
	printf '%s: error configuring %s\n' "$PROGNAME" "${STEM}" >&2
	exit 1
    fi
fi

echo "./configure --prefix=/tmp/mine --enable-maintainer-mode" >&2
if ! ./configure --prefix=/tmp/mine --enable-maintainer-mode
then
    printf '%s: error configuring %s\n' "$PROGNAME" "${STEM}" >&2
    exit 1
fi

echo "make -j2 all" >&2
if ! make -j2 all
then
    printf '%s: error configuring %s\n' "$PROGNAME" "${STEM}" >&2
    exit 1
fi

echo "make install" >&2
if ! make install
then
    printf '%s: error configuring %s\n' "$PROGNAME" "${STEM}" >&2
    exit 1
fi

exit 0

# Local Variables:
# mode: sh
# End:


src/ccmemory-standard-allocator.c
/*
  Part of: CCMemory
  Contents: implementation of standard memory allocator
  Date: May  5, 2018

  Abstract

	This module implements the standard memory allocator, relying on
	the standard  functions "malloc()", "realloc()",  "calloc()" and
	"free()".

  Copyright (C) 2018, 2020 Marco Maggi <mrc.mgg@gmail.com>

  This is free software; you  can redistribute it and/or modify it under
  the terms of the GNU Lesser General Public License as published by the
  Free Software  Foundation; either version  3.0 of the License,  or (at
  your option) any later version.

  This library  is distributed in the  hope that it will  be useful, but
  WITHOUT   ANY  WARRANTY;   without  even   the  implied   warranty  of
  MERCHANTABILITY  or FITNESS  FOR A  PARTICULAR PURPOSE.   See  the GNU
  Lesser General Public License for more details.

  You  should have  received a  copy of  the GNU  Lesser  General Public
  License along  with this library; if  not, write to  the Free Software
  Foundation, Inc.,  59 Temple Place,  Suite 330, Boston,  MA 02111-1307
  USA.
*/


/** --------------------------------------------------------------------
 ** Headers.
 ** ----------------------------------------------------------------- */

#include "ccmemory-internals.h"


/** --------------------------------------------------------------------
 ** Predefined memory allocator.
 ** ----------------------------------------------------------------- */

static void *
ccmem_standard_allocator_malloc (cce_destination_t L, ccmem_allocator_t const * const A CCLIB_UNUSED, size_t size)
{
  return cce_sys_malloc(L, size);
}

void *
ccmem_standard_allocator_realloc (cce_destination_t L, ccmem_allocator_t const * const A CCLIB_UNUSED, void * ptr, size_t newsize)
{
  return cce_sys_realloc(L, ptr, newsize);
}

void *
ccmem_standard_allocator_calloc (cce_destination_t L, ccmem_allocator_t const * const A CCLIB_UNUSED, size_t count, size_t eltsize)
{
  return cce_sys_calloc(L, count, eltsize);
}

void
ccmem_standard_allocator_free (ccmem_allocator_t const * const A CCLIB_UNUSED, void * ptr)
{
  free(ptr);
}

static ccmem_allocator_methods_t const ccmem_standard_allocator_methods_stru = {
  .malloc	= ccmem_standard_allocator_malloc,
  .realloc	= ccmem_standard_allocator_realloc,
  .calloc	= ccmem_standard_allocator_calloc,
  .free		= ccmem_standard_allocator_free
};

static ccmem_allocator_t const ccmem_standard_allocator_stru = {
  .methods	= &ccmem_standard_allocator_methods_stru
};

ccmem_allocator_t const * const ccmem_standard_allocator = &ccmem_standard_allocator_stru;

/* end of file */
